### ISSUE TITLE
Fastify v4 release coordination

### ISSUE THREAD

**Issue by mcollina** [fastify/fastify#3482]
This is an issue to coordinate the release of Fastify v4. It will be updated as soon as it progresses.

---

* [x] release new major of fast-json-stringify https://github.com/fastify/fast-json-stringify/releases/tag/v3.0.0
* [x] create type providers modules https://github.com/fastify/fastify/issues/3404
* [x] finish off all outstanding issues and PRs labeled v4: https://github.com/fastify/fastify/issues?q=is%3Aopen+is%3Aissue+label%3Av4.x and https://github.com/fastify/fastify/pulls?q=is%3Aopen+is%3Apr+label%3Av4.x
* [x] prepare release notes & blog post
* [x] check if anything needs to be done to support pino transports & pino pretty deprecation
* [x] prepare the list of modules to update
* [x] decide if we want to drop Node v12 support
* [x] drop Node v12 as well as v10 from everywhere https://github.com/fastify/fastify/pull/3532
* [x] sync fast-json-stringify next branch and release it (https://github.com/fastify/fast-json-stringify/tree/next)
* [x] implements https://github.com/fastify/fastify/issues/2852 by editing https://github.com/fastify/ajv-compiler ’s defaults `major` (ETA ~1h)
* [x] create a `fjs-compiler` module in the fastify’s org:
  * supersedes https://github.com/fastify/ajv-compiler/pull/45 (using the fast-json-stringify@next)
  * supersedes https://github.com/fastify/fastify/pull/3280
* [x] review/support https://github.com/fastify/fastify/pull/3546
* [x] remove `preParsing` hook from two fastify’s modules https://github.com/fastify/fastify/issues/3503#issuecomment-982531271
* [x] implement deprecation warning https://github.com/fastify/fastify/issues/3503
* [x] implement https://github.com/fastify/fastify/issues/3217
* [ ] Update all the core modules, for each one of them the following must be done:
  1. Open a PR updating Fastify to v4.0.0-rc.x
  2. Link the PR in the underlining list
  4. Once merged, release a new version of the module under the `next` dist-tag (`npm publish --tag next`)
  5. check `module@latest` is still correctly pointing to the old version with `npm show`
  6. tick the box in the underlining table
* [x] publish blog post
* [x] open issue at https://github.com/github/release-radar

---

## Modules to update

Once one it's done add the PR link next to it.

* [x] [`aws-lambda-fastify`](https://github.com/fastify/aws-lambda-fastify) - https://github.com/fastify/aws-lambda-fastify/pull/108
* [x] [`create-fastify`](https://github.com/fastify/create-fastify)
* [x] [`@fastify/accepts`](https://github.com/fastify/fastify-accepts) - https://github.com/fastify/fastify-accepts/pull/131
* [x] [`@fastify/accepts-serializer`](https://github.com/fastify/fastify-accepts-serializer) - https://github.com/fastify/fastify-accepts-serializer/pull/77
* [x] [`@fastify/auth`](https://github.com/fastify/fastify-auth) - https://github.com/fastify/fastify-auth/pull/154
* [x] [`@fastify/autoload`](https://github.com/fastify/fastify-autoload) - https://github.com/fastify/fastify-autoload/pull/239 - mostly not needed, the plugin just works.
* [x] [`fastify-awilix`](https://github.com/fastify/fastify-awilix) - https://github.com/fastify/fastify-awilix/pull/73
* [x] [`@fastify/basic-auth`](https://github.com/fastify/fastify-basic-auth) - https://github.com/fastify/fastify-basic-auth/pull/91
* [x] [`@fastify/bearer-auth`](https://github.com/fastify/fastify-bearer-auth) -https://github.com/fastify/fastify-bearer-auth/pull/134
* [x] [`@fastify/caching`](https://github.com/fastify/fastify-caching) - https://github.com/fastify/fastify-caching/pull/102
* [x] [`@fastify/circuit-breaker`](https://github.com/fastify/fastify-circuit-breaker) - https://github.com/fastify/fastify-circuit-breaker/pull/90
* [x] [`fastify-cli`](https://github.com/fastify/fastify-cli) - https://github.com/fastify/fastify-cli/pull/490
* [x] [`@fastify/compress`](https://github.com/fastify/fastify-compress) - https://github.com/fastify/fastify-compress/pull/224
* [x] [`@fastify/cookie`](https://github.com/fastify/fastify-cookie) - https://github.com/fastify/fastify-cookie/pull/181 (no changes needed)
* [x] [`@fastify/cors`](https://github.com/fastify/fastify-cors) - https://github.com/fastify/fastify-cors/pull/201
* [x] [`@fastify/csrf-protection`](https://github.com/fastify/fastify-csrf) - https://github.com/fastify/fastify-csrf/pull/89
* [x] [`@fastify/diagnostics-channel`](https://github.com/fastify/fastify-diagnostics-channel) - https://github.com/fastify/fastify-diagnostics-channel/pull/39 - not many changes needed, only in tests
* [ ] [`@fastify/elasticsearch`](https://github.com/fastify/fastify-elasticsearch) - https://github.com/fastify/fastify-elasticsearch/pull/77
* [x] [`@fastify/env`](https://github.com/fastify/fastify-env) - https://github.com/fastify/fastify-env/pull/144
* [x] [`@fastify/etag`](https://github.com/fastify/fastify-etag) - https://github.com/fastify/fastify-etag/pull/82
* [x] [`@fastify/express`](https://github.com/fastify/fastify-express) - https://github.com/fastify/fastify-express/pull/86
* [x] [`@fastify/flash`](https://github.com/fastify/fastify-flash) - https://github.com/fastify/fastify-flash/pull/71
* [x] [`@fastify/formbody`](https://github.com/fastify/fastify-formbody) - https://github.com/fastify/fastify-formbody/pull/111
* [x] [`@fastify/funky`](https://github.com/fastify/fastify-funky)
* [x] [`@fastify/helmet`](https://github.com/fastify/fastify-helmet) - https://github.com/fastify/fastify-helmet/pull/180
* [x] [`@fastify/http-proxy`](https://github.com/fastify/fastify-http-proxy) - https://github.com/fastify/fastify-http-proxy/pull/247
* [x] [`@fastify/jwt`](https://github.com/fastify/fastify-jwt) - https://github.com/fastify/fastify-jwt/pull/227 (not many change needed)
* [x] [`@fastify/leveldb`](https://github.com/fastify/fastify-leveldb) - https://github.com/fastify/fastify-leveldb/pull/93
* [x] [`@fastify/mysql`](https://github.com/fastify/fastify-mysql) - https://github.com/fastify/fastify-mysql/pull/92
* [x] [`@fastify/mongodb`](https://github.com/fastify/fastify-mongodb) - https://github.com/fastify/fastify-mongodb/pull/179
* [x] [`@fastify/multipart`](https://github.com/fastify/fastify-multipart) - https://github.com/fastify/fastify-multipart/pull/344 (no real change needed, only deprecations)
* [ ] ~[`@fastify/oauth2`](https://github.com/fastify/fastify-oauth2)~ skip because https://github.com/fastify/fastify-oauth2/issues/152
* [x] [`@fastify/passport`](https://github.com/fastify/fastify-passport) - https://github.com/fastify/fastify-passport/pull/552
* [x] [`@fastify/postgres`](https://github.com/fastify/fastify-postgres) - https://github.com/fastify/fastify-postgres/pull/131
* [x] [`@fastify/rate-limit`](https://github.com/fastify/fastify-rate-limit) - https://github.com/fastify/fastify-rate-limit/pull/239
* [x] [`@fastify/request-context`](https://github.com/fastify/fastify-request-context) - https://github.com/fastify/fastify-request-context/pull/114
* [x] [`@fastify/response-validation`](https://github.com/fastify/fastify-response-validation)
* [ ] ~[`@fastify/nextjs`](https://github.com/fastify/fastify-nextjs)~ - blocked by https://github.com/fastify/fastify-nextjs/pull/534
* [x] [`@fastify/redis`](https://github.com/fastify/fastify-redis) - https://github.com/fastify/fastify-redis/pull/147
* [x] [`@fastify/reply-from`](https://github.com/fastify/fastify-reply-from) - https://github.com/fastify/fastify-reply-from/pull/250
* [x] [`@fastify/routes`](https://github.com/fastify/fastify-routes) - https://github.com/fastify/fastify-routes/pull/66
* [ ] ~[`@fastify/routes-stats`](https://github.com/fastify/fastify-routes-stats)~ blocked by https://github.com/fastify/fastify-routes-stats/issues/76
* [x] [`fastify-schedule`](https://github.com/fastify/fastify-schedule) - https://github.com/fastify/fastify-schedule/pull/82
* [x] [`@fastify/sensible`](https://github.com/fastify/fastify-sensible) - https://github.com/fastify/fastify-sensible/pull/109
* [x] [`@fastify/secure-session`](https://github.com/fastify/fastify-secure-session) - https://github.com/fastify/fastify-secure-session/pull/130
* [x] [`@fastify/session`](https://github.com/fastify/session) - https://github.com/fastify/session/pull/89
* [x] [`@fastify/static`](https://github.com/fastify/fastify-static) - https://github.com/fastify/fastify-static/pull/292
* [x] [`@fastify/swagger`](https://github.com/fastify/fastify-swagger) - https://github.com/fastify/fastify-swagger/pull/599
* [x] [`@fastify/websocket`](https://github.com/fastify/fastify-websocket) - https://github.com/fastify/fastify-websocket/pull/197
* [x] [`@fastify/url-data`](https://github.com/fastify/fastify-url-data) - https://github.com/fastify/fastify-url-data/pull/104 (work as is, no need to update urgently)
* [x] [`middie`](https://github.com/fastify/middie) - https://github.com/fastify/middie/pull/144
* [x] [`point-of-view`](https://github.com/fastify/point-of-view) - https://github.com/fastify/point-of-view/pull/309
* [x] [`under-pressure`](https://github.com/fastify/under-pressure) - https://github.com/fastify/under-pressure/pull/162

**Comment by zekth** [comment]
Shouldn't we make a proper documentation review before landing v4? I know there were several corrections regarding phrasing / misleading sentences.

**Comment by mcollina** [comment]
> Shouldn't we make a proper documentation review before landing v4? I know there were several corrections regarding phrasing / misleading sentences.

I don't think it's feasible. We need to ship a v4 release to get on the latest Ajv and Pino bandwagon, otherwise we are keeping ourselves further behind. Moreover there are quite a few improvements sitting in the next branch. I would also like to phase out Node v10 support asap (and possibly v12).

**Comment by jsumners** [comment]
Agreed that docs shouldn't block (despite the mess). Just drop 12 and below. Our release cycle is too slow to carry that forward.

**Comment by jsumners** [comment]
> [ ] prepare the list of modules to update

I think we decided to move to `@fastify/`. So the list is: all of them. Which means I need help with https://github.com/fastify/deprecate-modules

Here's the thing: I am surfing the edge of burnout. I'm doing what I can to stay in the game at work, but OSS in my free time is really hard right now (I suspect the same is true of several of us). There's a mountain of things I want to get done here, in Pino, `ldapjs`, and my personal projects, but I just can't spend so much time on it.

**Comment by mcollina** [comment]
Thanks for the transparency @jsumners! I think a lot of us are in the same mindset at the minute.

**Comment by Fdawgs** [comment]
+1 on dropping support for node 10 and 12. Can mass remove them from CI workflows etc. if/when needed.

**Comment by climba03003** [comment]
Do anyone in @fastify/typescript have time to investigate on merging generic params for `FastifyInstance` and the others? 
If nobody have time to do so, I think it should be left for the next main version.

**Comment by darkgl0w** [comment]
I am +1 on dropping support for Node.js 10 and 12.

**Comment by Ethan-Arrowood** [comment]
@climba03003 do you have something I can look at for 

> merging generic params for FastifyInstance

**Comment by climba03003** [comment]
> @climba03003 do you have something I can look at for
> 
> > merging generic params for FastifyInstance

For example

https://github.com/fastify/fastify/blob/08877f173c3c3cd9e3883c97ca3285736d33832f/fastify.d.ts#L29-L34

should be changed to

```ts
// name can be changed
interface CommonServerInterface {
  Server = unknown, 
  Request = unknown, 
  Reply = unknown, 
  Logger = unknown 
}

declare function fastify<CommonServerInterface>(opts: FastifyHttp2SecureOptions<CommonServerInterface>): FastifyInstance<CommonServerInterface> & PromiseLike<FastifyInstance<CommonServerInterface>>
```
I don't know if it is possible, but a single interface here provide better DX if user want to override `Logger` only.

**Comment by Ethan-Arrowood** [comment]
Ahh yeah i've tried that once before and couldn't get it to work; maybe I'll try it again (but other folks should feel free to give it a shot + we don't have to block v4 for it)

**Comment by rocha-vitor** [comment]
Fastify is amazing, I'm using in some projects where I work, but I still can't implement in others projects I would like. The reason why is about the types. I really support that in the next version, has a comprehensively support for all features in fastify with consistent types, because somethings is very generics.

extra: the newer version of ajv has a support to serialization and seems be very fast.

**Comment by jsumners** [comment]
> [Better support for Typescript] Fastify is amazing, I'm using in some projects where I work, but I still can't implement in others projects I would like. The reason why is about the types. I really support that in the next version, has a comprehensively support for all features in fastify with consistent types, because somethings is very generics.
> 
> extra: the newer version of ajv has a support to serialization and seems be very fast.

If you have a new issue, please a new issue.

**Comment by mcollina** [comment]
I have successfully merged main into next. I will soon create a v3.x branch and move next to main.

**Comment by jsumners** [comment]
We leverage the GitHub releases feature to publish release notes.

**Comment by luisorbaiceta** [comment]
I'm aware. But I feel like there's people that miss those and instead just rely on our website documentation. 

The section wouldn't be something too extense, just the news about a new major version and a short description of the new features (and fixes?).

I see that there was a blog post planned too. We could also link it from the site.

**Comment by Fdawgs** [comment]
Like the [V3 migration guide](https://github.com/fastify/fastify/blob/main/docs/Guides/Migration-Guide-V3.md) but for V4?

**Comment by luisorbaiceta** [comment]
> Like the [V3 migration guide](https://github.com/fastify/fastify/blob/main/docs/Guides/Migration-Guide-V3.md) but for V4?

That could be added too!

So apart from the github releases feature to publish release notes we could:

* Make a short banner/section in our homepage to announce the release (this could include a link to the above mentioned blog post). This would be a short list with the key new improvements with a short description.
* Migration guide doc page (Under a new migrations section maybe?)

**Comment by jsumners** [comment]
Are we going to deprecate the `fastify-*` modules and replace them with `@fastify/*` versions? If so, I have some ideas on how to automate more of https://github.com/fastify/deprecate-modules

**Comment by mcollina** [comment]
I've been thinking about the renaming and I'm conflicted.

I see three options:

1. move them to Fastify v4 and rename, bumping major only once. Deprecate all the previous one.

2. rename all of them to `@fastify/*` and bump major. Then, upgrade to Fastify v4 and bump major again.

3. move them to Fastify v4 and rename, bumping major. Deprecate the previous ones only after v3 goes out of LTS.

There is a problem with 1) because there will be a window in which we are still supporting Fastify v3 and we will be issuing deprecation notices for a lot of modules without a replacement.

What do you think?

**Comment by jsumners** [comment]
> 2\. rename all of them to `@fastify/*` and bump major. Then, upgrade to Fastify v4 and bump major again.

This is what I was thinking would be done. Basically, the current majors of all of the `fastify-*` would be deprecated, even on `fastify@3`. The new `@fastify/*` modules would be the current modules for `fastify@3`. This would be done _prior_ to the final push for `fasitfy@4`.

**Comment by mcollina** [comment]
We are in agreement, so let's get it going. You said you had some script?

**Comment by jsumners** [comment]
> We are in agreement, so let's get it going. You said you had some script?

https://github.com/fastify/deprecate-modules is the proof of concept that needs to be completed. We should be able to write a script to iterate the org's repos and build the input data file. I _might_ be able to get to it this weekend, but I am having oral surgery on 2022-02-24 and will be rather out of commission.

**Comment by mcollina** [comment]
There is no hurry.

**Comment by skimmilk6877** [comment]
Is there any plan rewrite by TS?

**Comment by Fdawgs** [comment]
Next major release of shared workflows will drop support for node 10 and 12 as part of this: https://github.com/fastify/workflows/pull/15

**Comment by climba03003** [comment]
Issues that will potentially fixed by `v4`.
#3133 #3329

**Comment by mcollina** [comment]
I plan to release v4 in May.

**Comment by climba03003** [comment]
Not stale, it is in progress.

**Comment by Fdawgs** [comment]
Stalebot getting inpatient 😆

**Comment by jsumners** [comment]
These will be new major versions. Do not forget to update the name in the module metadata.

**Comment by Fdawgs** [comment]
https://github.com/fastify/fastify-bankai is archived, so can be removed from the list.

**Comment by mcollina** [comment]
Hey folks, I've covered most of the plugins in the list. I plan to have Fastify v4 released between now and OpenJS World in Austin.

**Comment by mcollina** [comment]
Feel free to take a stab at the remaining ones.

**Comment by jsumners** [comment]
Amazing work. I know how tedious that must've been.

**Comment by Retro64** [comment]
Really looking forward for the 4.x release. Great work! @mcollina  One question: Is `fastify-plugin` intentionally missing in the list "modules to update"?

**Comment by mcollina** [comment]
It does not need an update.

**Comment by metcoder95** [comment]
Hi! Made a couple of PRs for [@fastify/caching](https://github.com/fastify/fastify-caching/pull/102) and [@fastify/elasticsearch](https://github.com/fastify/fastify-elasticsearch/pull/77), I'll try to open more during the day/week 🙂

**Comment by mcollina** [comment]
I think we are almost ready to ship. I plan to issue a release after or during the Collaborator Summit in Austin.

**Comment by jsumners** [comment]
The modules issue still has a few things that need to be resolved.

**Comment by mcollina** [comment]
I'm happy to ship without those for now

**Comment by josh-hemphill** [comment]
I commented a while back about changing defaults and doc examples to JTD by default instead of JSON Schema. Some of that lies with ajv-compiler, but the jsonShorthand default currently has to be disabled by the user to use JTD schemas with the ajv-compier, which isn't a great experience for new users who want to use JTD.
I saw there was mention of putting off some changes so that v4 can land with the dependency upgrades, so perhaps changing the default could be pushed to something like v4.1; someone using JSON Schema syntax may perceive it as a breaking API change, but it is only a single option that changes the schema interpretation in some specific cases, so perhaps it could be defended as a semver minor change if necessary.
Just want to make sure it doesn't get overlooked if it needs to be in v4.

**Comment by jsumners** [comment]
I do not think JTD is a sufficiently widely adopted standard across languages to make it the default. JSON Schema enables interoperability among many different ecosystems.

**Comment by mweberxyz** [comment]
Regarding "check if anything needs to be done to support pino transports & pino pretty deprecation" -- in my recent testing my application started throwing deprecation warnings like:

`(node:56488) [PINODEP008] PinoWarning: prettyPrint is deprecated, look at https://github.com/pinojs/pino-pretty for alternatives.`

As suggested [here](https://github.com/pinojs/pino-pretty#integration) the deprecation warnings go away by changing my config from:
`fastify({logger: { prettyPrint: true}})`
to
`fastify({logger: { transport: { targets: 'pino-pretty' }}})`

I am seeing no other problems with newest pino and pino-pretty otherwise. If this is the case, the only needed change would be a documentation update. I'd be happy to create a PR to update the logging documentation if no one has picked it up yet.

**Comment by mcollina** [comment]
Go for it! Thanks!
